### ISSUE TITLE
[JOSS Review] Web-based applications

### ISSUE THREAD

**Issue by FlorianK13** [RWTH-EBC/AixWeather#69]
This comment is part of https://github.com/openjournals/joss-reviews/issues/7344

First of all, I think it is great that you offer AixWeather as a web app at https://aixweather.eonerc.rwth-aachen.de/
Most users are probably not interested in running the code on their own and just want to transform or download data, which is essentially what you provide. Going the extra step by hosting your services publicly is great. 

On web-based apps, they write in the [joss review guidelines](https://joss.readthedocs.io/en/latest/submitting.html#a-note-on-web-based-software):
> A note on web-based software
> 
> Many web-based research tools are out of scope for JOSS due to a lack of modularity and challenges testing and maintaining the code. Web-based tools may be considered ‘in scope’ for JOSS, provided that they meet one or both of the following criteria: 1) they are are built around and expose a ‘core library’ through a web-based experience (e.g., R/[Shiny](https://www.rstudio.com/products/shiny/) applications) or 2) the web application demonstrates a high-level of rigor with respect to domain modeling and testing (e.g., adopts and implements a design pattern such as [MVC](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) using a framework such as [Django](https://www.djangoproject.com/)).
> Similar to other categories of submission to JOSS, it’s essential that any web-based tool can be tested easily by reviewers locally (i.e., on their local machine).

I suppose (and correct me if I'm wrong) that you are aiming at category 1 "core library". By splitting the web app code into [this](https://github.com/RWTH-EBC/AixWeather-WebApp) seperate repository one essential step is already done. However, to see AixWeather as a real "core library", I miss these two things:
* Clear installation instructions and a prominent reference to your pypi package
* Some sort of "Getting started" on your documentation page, that guides me through the core process of the package, with examplary code to import, transform, and export weather data

**Comment by MartinRaetz** [comment]
Thank you @FlorianK13.

I do think commit f305373022eaf83e7c0ac8f69f427f84a57c9b50 and d2d7337fe39d64cb2043881f93ac673fae71f463 will solve this issue. 

Please close the issue if things are clear now.

**Comment by FlorianK13** [comment]
We do agree on the “99% of the users will go for the trivial web app” – as mentioned above I like this web app a lot for its simplicity. It gives most users exactly what they want, ie the transformed weathr data. I also understand your points about maintenance and single source of truth.
For joss however the review focuses on the core software package, its functionalities, modularity, and reusability. A docs page is the core feature to assure reusability. Things that should be answered from the docs page:
-	What data do I get?
-	What are the different transformation processes the package offers?
-	AixWeather ensures data accuracy? What does that mean? In which functions does this happen? What does happen there? Especially for this sentence in your paper I cannot find references in your documentation: “including consideration of time zones, the time reference of the measurement, unit conversions, correct handling of incomplete data series and, where possible, the avoidance of interpolation-related smoothing.” 
-	What are typical use cases and workflows where the usage of AixWeather makes my live easier?

Ultimately if I want to reuse AixWeather as a software package in something that I build on my own, all these questions are important and I think they need to be answered. And an example python file is, in my opinion, not enough to achieve that.

**Comment by FlorianK13** [comment]
However, since I’m also rather new as a reviewer for joss, these requirements come from my own interpretation of the joss review docs. Maybe I’m wrong, so I would also be happy to include the other reviewer and/or the editor into this discussion.

**Comment by MartinRaetz** [comment]
Hey @FlorianK13,

thank you for specifying! 
I will think about how to integrate such information without generating something that needs regular maintenance. 

Here is a commit that would hopefully clarify some of your points: 1234a63c2fc819cf8eb9753a402d65337377222c

Now some explanations regarding your questions:

You asked what data the user gets. Well, I would describe it with "Weather data" for building energy simulations. (I have clarified this in the commit). The input and output formats, as well as the transformation processes, are shown in the figure. What data points these are exactly, will always depend on the desired format and the availability of the data. (I have also clarified this.) The same applies to the transformation processes, if the unit of a data point is "wrong" a transformation to the desired unit is done. Otherwise, no unit transformation is done. So this will always depend on the input data and the chosen output format. 

Speaking about the use cases. I do think it's building energy simulations. I mean you might use the weather data for whatever you want to do, though the tool is designed to deliver weather data for building energy simulations. 

Regarding the "accuracy", I have adapted the formulation in the paper to make it clearer. I meant, that accuracy is ensured through the consideration of the stated points. Also, these points are not super complex functions. Consideration of time zones just means, that the tool considers (and knows) the time zone of the raw input data and shifts the data accordingly to the desired time zone in the output format. Same simplicity for the units, respecting incomplete data without applying wrong interpolation schemes. So these points are important, and actually often done wrong, but intrinsically trivial.
The interpolation-related smoothing is explained in the figure. 

If you want to reuse AixWeathers internal functions, you will need to dive into the code. These functions are under constant extension and improvement. Not only because the tool is supposed to grow and support more data formats, but also cause input formats or output formats might change and the tool needs to be updated to these changes. If you just want to use AixWeathers functionality to get weather data, I think the example file is exactly what you need. You can copy it to your code and you're done.

**Comment by FlorianK13** [comment]
I think our discussion is going in circles. As I see it, the primary goal of AixWeather is the provision of the web service. Creating and maintaining a large documentation page is lots of work, which contrasts the primary goal of providing the web service, as it shifts focus away from work that benefits the web service.

JOSS however is a journal about open source software packages, from and for the research community. If other researchers want to use the core software of AixWeather, they need to know in detail what it provides. Hence, JOSS specifically requests a review of the available documentation. In my opinion, the current docs of AixWeather fail this part of the review.

What I expect from a docs page:
- I want to understand, as quickly as possible, what the provided software can do for me. Is it really what I need? Here it would be nice to have some kind of "statement of need" as the first visible thing at https://rwth-ebc.github.io/AixWeather//main/docs/index.html with the information that I can get DWD weather data and export it in format a, b, c.
- I now see that this is the software I need. How do I integrate it in my code? I maybe want to import from source A, do some of the transformations, and then export to format B. How do I do that? What transformations can I apply, which sources do exist, which exports? Some sort of manual with code and references to the API are needed here.

In the end, you are the maintainer of this software, so you now best how your potential users can use it and what questions might come up for them. 

I also checked the last five publications on JOSS and their docs pages: [snk](https://snk.wytamma.com/managing_workflows/), [BrightEyes-MCS](https://github.com/VicidominiLab/BrightEyes-MCS/wiki#input--output-connections), [Phylox](https://phylox.readthedocs.io/en/latest/index.html), [PhenoFeatureFinder](https://phenofeaturefinder.readthedocs.io/en/latest/index.html), [lpjmlkit-1](https://pik-piam.r-universe.dev/lpjmlkit), [lpjmlkit-2](https://pik-piam.r-universe.dev/lpjmlkit/doc/manual.html)

**Comment by lymereJ** [comment]
Sorry for including myself in the discussion. I left some preliminary review comments [here](https://github.com/openjournals/joss-reviews/issues/7344) and I see that some of them are echoed here by @FlorianK13. I agree that the documentation should be expanded as mentioned in the comment above. There is always a possibility that funding for hosting the web application dries up. In that case, it is essential to have a good documentation so people can use the tool offline.

**Comment by MartinRaetz** [comment]
Thank you @FlorianK13 and @lymereJ, 

- [ ] @MartinRaetz: I will have another look at the documentation, with your suggestions and thoughts I hope to come up with something that covers both needs. However, I might need some time to do a deeper dive.

What are your thoughts on shutting down the GitHub docs page and focusing documentation only on the readme (and example file)? The external docs page always comes with quite some pain to maintain, and everything except for the "aixweather" tab with the modules overview, is (or can) be included in the Readme. Also, bear in mind that the docs page you see is only rendered for the current main branch, so changes I do now are not automatically mirrored online.
I also think, that since the figures are not rendering on the docs page, people might overlook the information in the "structure" figure if they decide to read the "readme" on the docs pages and not on the readme page. It also might deliver a wrong image of the tool. E.g. the tool is not meant to be used like e.g. Scikit-Learn where you have many functions that you can combine to your liking to your "own" script. AixWeather is and must be, a streamlined process, always running the same three functions in the same order. None of those internal transformation functions are designed to be used out of that process, the process is designed to choose the correct transformation on its own. And this is, in my opinion, the strength, since this will ensure that you will get the weather data correctly transformed to your desired output format from an often obscure input data format. 

I will also try to include these thoughts in the documentation.

**Comment by FlorianK13** [comment]
I'm fine with a documentation in the README

**Comment by lymereJ** [comment]
Sorry for the late reply, same.

**Comment by MartinRaetz** [comment]
Hey @lymereJ and @FlorianK13,

fortunately, we (mainly @FWuellhorst) could fix and improve the CI that creates the docs pages. It is now a lot more stable.

Additionally, I have improved the documentation. There is now a direct explanation of what the tool is and what to expect from it right at the start of the readme and docs pages, and again in the tutorial/example (@FlorianK13 called it statement of need). Also, in the "how to use" section more information is added and the user is forwarded to a tutorial/example available on the docs page. To enable you to check the updated docs pages I have temporarily linked the pages generated by the joss-branch. I will change this back to the main branch upon merging. 

When checking, please make sure to be in the https://github.com/RWTH-EBC/AixWeather/tree/4-joss_paper branch. 

Thank you for your patience, unfortunately, this review coincides with the last months of my PhD (which is not related to this open-source contribution). 

I will now have a look at the things @lymereJ mentioned in the joss-review thread. 

Thanks again for volunteering!

**Comment by FlorianK13** [comment]
The new docs can be accessed directly here, might be useful for review @MartinRaetz @lymereJ  https://rwth-ebc.github.io/AixWeather//4-joss_paper/docs/index.html#

**Comment by FlorianK13** [comment]
The work on the docs page looks good and my major concerns were adressed. The following points are only smaller things:
- I'm missing a link to the actual webapp in your [Readme](https://github.com/RWTH-EBC/AixWeather/tree/4-joss_paper?tab=readme-ov-file#webapp) 
- When you write about the [project classes](https://rwth-ebc.github.io/AixWeather//4-joss_paper/docs/examples/e1_pull_DWD_historical_to_all_output_formats.html) in the tutorial, you could link to the [API documentation](https://rwth-ebc.github.io/AixWeather//4-joss_paper/docs/code/aixweather.html#module-aixweather.project_class) of those classes
- I was struggling to get a second example running. Maybe you can extend the tutorial with one more project class, eg the DWD Forecast?

**Comment by lymereJ** [comment]
Thanks @MartinRaetz and @FlorianK13! This looks great.

On the main page of the documentation you have the following:
```
Clone the repo and simply install via:

pip install AixWeather

or

cd AixWeather
pip install -e .
```

I wonder if you shouldn't have the following:
```
Simply install via:

`pip install AixWeather`

Or, by cloning the repository:

`git clone https://github.com/RWTH-EBC/AixWeather.git AixWeather`
`cd AixWeather`
`pip install -e .`
```

**Comment by MartinRaetz** [comment]
Thank you @FlorianK13 and @lymereJ!

I have addressed your points in the last two commits. 

@FlorianK13: I have added additional "copy&paste"-snippets for the other project classes at the end of the tutorial. They work on my computer, let me know if you are having issues!

Thanks again for your time and suggestions!

Have a wonderful Christmas break!

**Comment by FlorianK13** [comment]
Looks good so far. I haven't managed to get the EPW and TRY example running. I think the reason is that the 
`definitions.ROOT_DIR` links to `.venv/lib/python3.12/site-packages` and not to the actual git repo if you have installed the package via pip. For me it would also be fine to have something like this in the docs:

"To parse EPW or TRY files, you need them locally. You can download example files from our repository at github here: INSERT LINK

Then parse them as follows:

project = ProjectClassEPW(
    path="path/to/your/epw/file"
    )
"

**Comment by MartinRaetz** [comment]
Hey @FlorianK13, 

just to verify, did you clone or download the repository and execute the example file from there?
To me the implementation of ROOT_DIR seems to be fine, it should take the location of the executed file as the origin and walk to the ROOT_DIR from there. This should be independent of the way the environment is installed. 

I have clarified the requirements as you suggested.

**Comment by FlorianK13** [comment]
Yes, it should work if people download the repo and install the software from there. But if people get software by installing it via pip, then the ROOT-DIR links to the `site-package` of the python environment - at least that is my assumption of this error.

Have you tried to install the software in a clean environment via pip and then run the example script from there?

**Comment by MartinRaetz** [comment]
Yes, downloading the repo to run the example is required (if you want to use files from the repo). I have clarified this in the comments. You may check the current version of the example. It will upload to the Docs Pages in the next 20 minutes or so.

**Comment by MartinRaetz** [comment]
Thank you @FlorianK13. 
 
 To speed up the process, here is what will be uploaded once the CI is finished:
  
For EPW and TRY we need to import the .epw or .dat file, e.g. from your local drive. The
Readme provides information on where you can download your own EPW and TRY files. For this
example, we use the test files provided in the tests folder (be sure to have the full
repository downloaded to have access to these files).
If you use your own files, you can specify the path to the file in the `path` argument.

Hope this clarifies the usage?
Thanks for suggesting!

**Comment by MartinRaetz** [comment]
Info: The CI is finished and the doc-pages are updated.

**Comment by MartinRaetz** [comment]
Since the JOSS review is completed and the paper published, this issue can be closed.
